ostd/src/mm/page_table/node/entry.rs

```diff
@@ -0,0 +1,147 @@
+    pub(in crate::mm) fn protect(&mut self, op: &mut impl FnMut(&mut PageProperty)) {
+        let prop = self.pte.prop();
+        let mut new_prop = prop;
+        op(&mut new_prop);
+        if prop != new_prop {
+            self.pte.set_prop(new_prop);
+            self.node.write_pte(self.idx, self.pte);
+        }
+    }
+
+    /// Replaces the entry with a new child.
+    ///
+    /// The old child is returned.
+    ///
+    /// The provided child must match with the level of the node.
+    pub(in crate::mm) fn replace(self, new_child: Child<E, C>) -> Child<E, C> {
+        // It should be ensured by the cursor.
+        #[cfg(debug_assertions)]
+        match &new_child {
+            Child::PageTable(_) => {
+                debug_assert!(self.node.level() > 1);
+            }
+            Child::Page(p, _) => {
+                debug_assert!(self.node.level() == p.level());
+                debug_assert!(self.node.is_tracked() == MapTrackingStatus::Tracked);
+            }
+            Child::Untracked(_, level, _) => {
+                debug_assert!(self.node.level() == *level);
+                debug_assert!(self.node.is_tracked() == MapTrackingStatus::Untracked);
+            }
+            Child::None => {}
+        }
+
+        // SAFETY: The entry structure represents an existent entry with the
+        // right node information. The old PTE is overwritten by the new child
+        // so that it is not used anymore.
+        let old_child =
+            unsafe { Child::from_pte(self.pte, self.node.level(), self.node.is_tracked()) };
+
+        if old_child.is_none() && !new_child.is_none() {
+            *self.node.nr_children_mut() += 1;
+        } else if !old_child.is_none() && new_child.is_none() {
+            *self.node.nr_children_mut() -= 1;
+        }
+
+        self.node.write_pte(self.idx, new_child.into_pte());
+
+        old_child
+    }
+
+    /// Splits the untracked huge page to smaller pages.
+    ///
+    /// The new child page table node is returned.
+    ///
+    /// This entry should be an untracked huge page.
+    pub(in crate::mm) fn split_untracked_huge(self) -> PageTableNode<E, C> {
+        // These should be ensured by the cursor.
+        debug_assert!(self.node.level() > 1);
+
+        let pa = self.pte.paddr();
+        let level = self.node.level();
+        let prop = self.pte.prop();
+
+        let mut new_page = PageTableNode::<E, C>::alloc(level - 1, MapTrackingStatus::Untracked);
+        for i in 0..nr_subpage_per_huge::<C>() {
+            let small_pa = pa + i * page_size::<C>(level - 1);
+            let _ = new_page
+                .entry(i)
+                .replace(Child::Untracked(small_pa, level - 1, prop));
+        }
+
+        let _ = self.replace(Child::PageTable(new_page.clone_raw()));
+
+        new_page
+    }
+
+    pub(super) fn new_at(node: &'a mut PageTableNode<E, C>, idx: usize) -> Self {
+        debug_assert!(idx < nr_subpage_per_huge::<C>());
```

Comment:
The three methods `new_at`/`read_pte`/`write_pte` can only be accessed in the node module. So the satisfiability is contained. My concern is that if we mark it unsafe there would be many trivially true unsafe blocks. I think using debug assertions to mark such safety preconditions are just enough. Can you provide more insights about how should we deal with such cases?

Reply:
> The three methods `new_at`/`read_pte`/`write_pte` can only be accessed in the node module. So the satisfiability is contained.

Yes, in theory it's fine to mark all internal methods as _not_ unsafe, and since they're not public, our code is still sound with respect to the public APIs.

So it's more of a readability and maintainability issue. 

> My concern is that if we mark it unsafe there would be many trivially true unsafe blocks.

First, I'm not sure why you say "many". Currently, I believe `read_pte` is called only once (when constructing `Entry`) and `write_pte` is called only twice (when operating on `Entry`), and both are definitely not "many".

Second, I would argue that if some unsafe methods are called many times with the same trivially true safety comments, this just signals that our code needs refactoring.

For example, if `write_pte` needs to be called 10 times when operating on `Entry`, I would suggest that a better strategy is to abstract a safe method to do this:
```rust
impl Entry {
    fn write_pte(&mut self, something: Something) {
        // SAFETY: `self.index` is in bounds, guaranteed by the safety conditions of `Entry::new_at`.
        unsafe { self.node.write_pte(self.index, something) };
    }
}
```

(Actually, the introduction of `Entry` did a good job and helps the above argument. As you can see, before this PR, `node.read_pte`/`node.write_pte` can be called directly and the caller has to make sure the index is in bounds, but now after `Entry` is created, the index of `Entry` is always in bounds, so we can easily create safe wrappers).

> I think using debug assertions to mark such safety preconditions are just enough.

At least to me, the current implementation does not seem very reasonable.

For example, the implementation of `read_pte` is:
```rust
fn read_pte(&self, idx: usize) -> E {
    // ...

    // SAFETY: the index is within the bound and PTE is plain-old-data.
    unsafe { ptr.add(idx).read() }
}
```

The statement "the index is within the bound" does not make much sense because `idx` is just an (unchecked) input argument. The code just says that some preconditions need to be _upheld_, but the comments say that the precondition is _already met_, they just contradict each other.

So my point is: the correct, Rust-style way to express that some conditions must be upheld, is to mark the function as unsafe. Again, if they create duplicates, try to find a better way to wrap these unsafe methods.

Reply:
> Second, I would argue that if some unsafe methods are called many times with the same trivially true safety comments, this just signals that our code needs refactoring.
> 
> For example, if `write_pte` needs to be called 10 times when operating on `Entry`, I would suggest that a better strategy is to abstract a safe method to do this:
> 
> ```rust
> impl Entry {
>     fn write_pte(&mut self, something: Something) {
>         // SAFETY: `self.index` is in bounds, guaranteed by the safety conditions of `Entry::new_at`.
>         unsafe { self.node.write_pte(self.index, something) };
>     }
> }
> ```

I also note that this is not currently the case. I believe the validity of the PTE passed to `write_pte` is also another safety condition that is just never mentioned in the current implementation.

For example, I expect that the reasons why `write_pte` is sound in `protect` / `replace` are actually _different_. So there is no duplication.

```rust
    /// Operates on the mapping properties of the entry.
    pub(in crate::mm) fn protect(&mut self, op: &mut impl FnMut(&mut PageProperty)) {
        // ...        

        // SAFETY:
        // 1. The index is inbounds.
        // 2. We replace the PTE with a new one, which differs only in `PageProperty`, so it is
        //    still compatible with the current page table node.
        unsafe { self.node.write_pte(self.idx, self.pte) };
    }

    pub(in crate::mm) fn replace(self, new_child: Child<E, C>) -> Child<E, C> {
        assert!(new_child.is_compatible(self.node.level(), self.node.is_tracked()));

        // ...

        // SAFETY:
        // 1. The index is in bounds.
        // 2. The new PTE is compatible with the page table node, as asserted above.
        self.node.write_pte(self.idx, new_child.into_pte());

        old_child
    }
```

Well, I'm not sure if you think it's "obvious", but I think having these comments makes our implementation clearer.

Reply:
Thanks very much for your explanation in depth. I have accepted your suggestion.